### PR TITLE
added toolbox item and toolbox item collection implementation

### LINKED ISSUE #478: Port Toolbox related components from framework to core
- [x] ToolboxItem
 - [x] IToolboxUser
 - [x] IToolboxService
 - [x] ToolboxItemCollection
 - [x] ToolboxItemCreatorCallback

Related requests: #221 #469

### PR DESCRIPTION
**Pull request by JuditRose**
Fixes #478 
Contributes to #221, #469 
Implementations are copied over from here: https://github.com/dotnet/corefx/commit/fc7b99477d4937b8c31ea5a9e647ec7578139328

### CODE REVIEW

**Comment by JuditRose** [comment]
Unit tests are coming later, I just wanted to open the PR to start the conversation :)

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.cs [hunk 1]
```diff
@@ -709,7 +709,7 @@ internal static void OpenFormsInternalRemove(Form form)
             }
         }
 
-        internal static bool UseVisualStyles {
+        public static bool UseVisualStyles {
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/Application.cs]: This was a bit tricky...

I needed to move CommonUnsafeNativeMethods to the Common so components in S.W.F.Design could use it. 
CommonUnsafeNativeMethods is trying to access UseVisualStyles from Application.
Application is a public class, so it is visible from S.W.F.Design, BUT UseVisualStyles was internal, so CommonUnsafeNativeMethods couldn't access if from S.W.F.Design. 

As Application is public, moving it to common will cause issues in S.W.F.Design because 
1. S.W.F.Design includes all public classes from S.W.F.
2. S.W.F. and S.W.F.Design both would include Application from the common folder 
1 + 2 => Application would be duplicated in S.W.F.Design. Which IMHO is not really a good idea, even if I could resolve the dependencies with direct route definitions.
So instead of moving Application class to the Common folder I made UseVisualStyles public. 
I am open for better solutions, I couldn't come up with any better.

File: src/System.Windows.Forms/src/System.Windows.Forms.csproj [hunk 2]
```diff
@@ -21,7 +21,20 @@
...
 
   <ItemGroup>
+    <Compile Include="..\..\Common\src\misc\CommonUnsafeNativeMethods.cs" Link="misc\CommonUnsafeNativeMethods.cs" />
+    <Compile Include="..\..\Common\src\misc\CompModSwitches.cs" Link="misc\CompModSwitches.cs" />
+    <Compile Include="..\..\Common\src\misc\DebugHandleTracker.cs" Link="misc\DebugHandleTracker.cs" />
+    <Compile Include="..\..\Common\src\misc\DpiHelper.cs" Link="misc\DpiHelper.cs" />
+    <Compile Include="..\..\Common\src\misc\ExternDll.cs" Link="misc\ExternDll.cs" />
+    <Compile Include="..\..\Common\src\misc\HandleCollector.cs" Link="misc\HandleCollector.cs" />
     <Compile Include="..\..\Common\src\ApiHelper.cs" Link="Common\ApiHelper.cs" />
+    <!--<Compile Include="..\..\Common\src\ClientUtils.cs" Link="Common\ClientUtils.cs" />-->
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System.Windows.Forms.csproj]: I will fix the csproj files after #439 is merged, so I can resolve the dependencies which are present in both changeset

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.cs (hunk 1 above)
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/Application.cs]: How about conditionally compile CommonUnsafeNativeMethods' method that uses `UseVisualStyle` property only when that file is referenced from S.W.F.csproj - `#if WINFORMS_NAMESPACE`

File: src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj [hunk 3]
```diff
@@ -21,4 +25,18 @@
     <ProjectReference Include="..\..\System.Windows.Forms\src\System.Windows.Forms.csproj" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Compile Include="..\..\Common\src\misc\CommonUnsafeNativeMethods.cs" Link="misc\CommonUnsafeNativeMethods.cs" />
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj]: BTW, we have `Unsafe` PInvoke declarations in the Classic Framework because we explicitly disabled some security checks on these methods - https://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/UnsafeNativeMethods.cs,172, see [SuppressUnmanagedCodeSecurity](https://docs.microsoft.com/en-us/dotnet/api/system.security.suppressunmanagedcodesecurityattribute?view=netframework-4.7.2) attribute because these methods are internal and we invoke them from the safe context or [Demand ](https://docs.microsoft.com/en-us/dotnet/api/system.security.permissionset.demand?view=netframework-4.7.2) appropriate permissions explicitly before calling it. These attribute is not applicable to Core. So there is no need to keep P/Invoke declarations in different files, all can be named `NativeMethods` but, this is outside of scope of this change.

**Comment by JuditRose** [comment]
toolbox

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.cs (hunk 1 above)
Comment (zsd4yr) [reply src/System.Windows.Forms/src/System/Windows/Forms/Application.cs]: Is it possible to move the method(s) that require UseVisualStyles out of CommonUnsafeNativeMethods and into a file that is specific to System.Windows.Forms? UnsafeNativeMethods.cs for example

File: src/System.Windows.Forms/src/System.Windows.Forms.csproj (hunk 2 above)
Comment (zsd4yr) [reply src/System.Windows.Forms/src/System.Windows.Forms.csproj]: #439 is merged 😄 have these been re-included?

**Comment by zsd4yr** [comment]
It looks like you're failing on the throw not implemented exceptions because those are no longer the case 😉

**Comment by JuditRose** [comment]
> It looks like you're failing on the throw not implemented exceptions because those are no longer the case 😉

I know! haha... this was the plan :) I am fixing those test cases and adding some more today

File: src/System.Windows.Forms.Design/tests/System/Drawing/Design/ToolboxItemCollectionTests.cs [hunk 4]
```diff
@@ -7,18 +7,24 @@
...
-        {	
-            ToolboxItem[]  tools = { new ToolboxItem(), new ToolboxItem() };	
-
-             return new ToolboxItemCollection(tools);	
-        }	
-
+    {
         [Fact]
-        public void TestToolboxItemCollection_Throw_NotImplemented_Exception()
+        public void TestToolboxItemCollection_Creation()
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/tests/System/Drawing/Design/ToolboxItemCollectionTests.cs]:
```suggestion
        public void ToolboxItemCollection_Creation()
```
Should be named after the class it tests

Comment (zsd4yr) [reply src/System.Windows.Forms.Design/tests/System/Drawing/Design/ToolboxItemCollectionTests.cs]: Ah a lot of these seem to be doing that. Would you mind changing them? Not sure why this was not caught earlier

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.cs (hunk 1 above)
Comment (JuditRose) [reply src/System.Windows.Forms/src/System/Windows/Forms/Application.cs]: `UseVisualStyles` returns a value of a static field existing in `Application`, and `UnsafeNativeMethods` (in a public static method called `Activate`) contains an if statement in the logic `if (Application.UseVisualStyles ) { .... }` . I don't see how `UseVisualStyles` could be copied over to `UnsafeNativeMethods`.

File: src/System.Windows.Forms/src/System.Windows.Forms.csproj (hunk 2 above)
Comment (JuditRose) [reply src/System.Windows.Forms/src/System.Windows.Forms.csproj]: Nope, just reordered alphabetically

File: src/System.Windows.Forms.Design/src/System/Drawing/Design/ToolboxItem.cs [hunk 5]
```diff
@@ -1060,7 +1060,7 @@ public override void Add(object key, object value)
 
             private void CheckSerializable(object value)
             {
-                if (value != null && !value.GetType().IsSerializable)
+                if (value != null && !(value.GetType().IsSerializable || value is ISerializable))
```
[the remark this answers is not in the document]
Comment (JuditRose) [reply src/System.Windows.Forms.Design/src/System/Drawing/Design/ToolboxItem.cs]: No, they are not equivalent. (If they were, this change would be irrelevant and unnecessary)
Every `ToolBoxItem` instantiation calls the method `Initialize`. 
In `Initialize` there is a variable defined:  `AssemblyName assemblyName = type.Assembly.GetName(true);`. 
The value of the `assemblyName` variable eventually gets set as a value of a member: `AssemblyName = assemblyName;`. 
During setting the value of `AssemblyName` there is a call to the `CheckSerializable` method which will end up in _this_ line checking if the type `AssemblyName` is serializable or not. (This happens every time, regardless of what was the type passed in the constructor for the ToolBoxItem)
If not, it throws an `ArgumentException` and the object won't get initialized. 
`IsSerializable` is a member on `Type` and for `AsseblyName.getType()` it returns false, although `AsseblyName` implements the ISerializable interface.  
(https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/serialization/ says `In custom serialization, you can specify exactly which objects will be serialized and how it will be done. The class must be marked SerializableAttribute and implement the ISerializable interface.`)
SO: `if (value != null && !value.GetType().IsSerializable)` for `AssemblyName` class will always be false and no `ToolBoxItem` can be instantiated/initialized ever.
(`AssemblyName` is a public sealed class in System.Reflection, which I wasn't comfortable changing)
There are a few possible solutions to enable instantiation, for me this seemed the most reasonable, but I am open for alternatives.

File: src/System.Windows.Forms.Design/tests/System/Drawing/Design/ToolboxItemCollectionTests.cs (hunk 4 above)
Comment (JuditRose) [reply src/System.Windows.Forms.Design/tests/System/Drawing/Design/ToolboxItemCollectionTests.cs]: Done

File: src/System.Windows.Forms.Design/src/System/Drawing/Design/ToolboxItem.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (zsd4yr) [reply src/System.Windows.Forms.Design/src/System/Drawing/Design/ToolboxItem.cs]: Ah so then should it only be value is ISerializable instead of the or?

[the remark this answers is not in the document]
Comment (JuditRose) [reply src/System.Windows.Forms.Design/src/System/Drawing/Design/ToolboxItem.cs]: I can definitely change that. What is the relation between `IsSerializable` and `ISerializable`? Do we prefer `ISerializable` here over `IsSerializable`? It could have been a bug before, I can imagine that.

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.cs (hunk 1 above)
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/Application.cs]: You could conditionally compile this method out of the file:
#if WINFORMS_NAMESPACE
public static IntPtr Activate()
<snip>
#endif

File: src/System.Windows.Forms.Design/src/System/Drawing/Design/ToolboxItem.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms.Design/src/System/Drawing/Design/ToolboxItem.cs]: When used in the VS, Toolbox items should be serializable (IsSerializable==true, both implementing ISerializable and having the attribute) because they are stored as strings when passing around and when initializing VS Toolbox. I would suggest replacing the AssemblyName-typed property with a string property (fully qualified assembly name). However I don't know how to handle such breaking change, perhaps @ericstj could advice. 
Or we can ignore serialization requirement for toolbox items for now because there is no toolbox to put them in at this point.

File: src/System.Windows.Forms.Design/src/Resources/SR.resx [hunk 6]
```diff
@@ -114,11 +173,38 @@
...
+  <data name="ToolboxItemLocked" xml:space="preserve">
+    <value>Toolbox item cannot be modified.</value>
+  </data>
+  <data name="ToolboxItemValueNotSerializable" xml:space="preserve">
+    <value>Data type {0} is not serializable. Items added to a property dictionary must be serializable.</value>
+  </data>
+  <data name="UnsafeNativeMethodsNotImplemented" xml:space="preserve">
+    <value>Not implemented.</value>
+  </data>
 </root>
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms.Design/src/Resources/SR.resx]: Interesting -we have editorconfig in this repo which forces `\r\n` in the end of the file. But you probably edited this file in the resx editor, and that editor did not respect our requirements? So my diff diff shows this:
image

```suggestion
</root>

```

File: src/System.Windows.Forms.Design/src/Resources/SR.resx [hunk 7]
```diff
@@ -1,5 +1,64 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <root>
+  <!--
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms.Design/src/Resources/SR.resx]: please remove the header

File: src/System.Windows.Forms.Design/src/Resources/xlf/SR.cs.xlf [hunk 8]
```diff
@@ -107,6 +127,31 @@
...
+        <note />
+      </trans-unit>
+      <trans-unit id="UnsafeNativeMethodsNotImplemented">
+        <source>Not implemented.</source>
+        <target state="new">Not implemented.</target>
+        <note />
+      </trans-unit>
     </body>
   </file>
 </xliff>
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms.Design/src/Resources/xlf/SR.cs.xlf]: Same bad line ending here. Perhaps if you open these files in xml editor in VS and then save, they would get cleaned up?

File: src/System.Windows.Forms.Design/src/System/Drawing/Design/ToolboxItem.cs [hunk 9]
```diff
@@ -7,6 +7,11 @@
 using System.ComponentModel;
 using System.ComponentModel.Design;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms.Design/src/System/Drawing/Design/ToolboxItem.cs]: These are not in alphabetical order

File: src/System.Windows.Forms.Design/src/System/Drawing/Design/ToolboxItem.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (ericstj) [reply src/System.Windows.Forms.Design/src/System/Drawing/Design/ToolboxItem.cs]: Returning true from Type.IsSerializable and implementing ISerializeable are different things.  Implementing ISerializable is just an implementation detail about how a type may decide to serialize itself.  To be Type.IsSerializable it still needs to set the attribute.

AssemblyName is not serilzeable at all on .NETCore.  Even though it implements ISerializable it [throws](https://github.com/dotnet/coreclr/blob/fd3afcaf0dc0ace3f84d0412c128f4e64c419105/src/System.Private.CoreLib/src/System/Reflection/AssemblyName.cs#L268-L271) PlatformNotSupportedException (likely just for API compatibility so that it doesn't fail JIT with missing method).  I believe the omission of serializability for this type was intentional due to it containing a CultureInfo, but @ViktorHofer may know better.  You shouldn't be relying on AssemblyName implementing ISerializable as any sort of indication of serializability.  Instead either special-case AssemblyName in the calling code to avoid trying to serialize it, or make a request on CoreCLR to enable serialization of AssemblyName if it's absolutely impossible to handle at this level.

[the remark this answers is not in the document]
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms.Design/src/System/Drawing/Design/ToolboxItem.cs]: We don't need to serialize ToolBoxItems at this point, there are no AppDomains to pass them between, and we will not be binary-serializing them when communicating with the out of proc designer. So we don't need the `CheckSerializable` method at all.

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.cs (hunk 1 above)
Comment (JuditRose) [reply src/System.Windows.Forms/src/System/Windows/Forms/Application.cs]: Is this better to do that instead of just make UseVisualStyles public?

File: src/System.Windows.Forms.Design/src/Resources/SR.resx (hunk 6 above)
Comment (JuditRose) [reply src/System.Windows.Forms.Design/src/Resources/SR.resx]: I edited this in VS, and whatever default editor it uses for resx files. Is there a setting that I need to change maybe?

File: src/System.Windows.Forms.Design/src/Resources/SR.resx (hunk 7 above)
Comment (JuditRose) [reply src/System.Windows.Forms.Design/src/Resources/SR.resx]: Same here, this one is added by VS automatically, is there a setting to turn it off?

Comment (ViktorHofer) [reply src/System.Windows.Forms.Design/src/Resources/SR.resx]: I don't think so. In corefx we always remove it manually afterwards.

File: src/System.Windows.Forms.Design/src/Resources/xlf/SR.cs.xlf (hunk 8 above)
Comment (JuditRose) [reply src/System.Windows.Forms.Design/src/Resources/xlf/SR.cs.xlf]: This change again has been automatically generated by VS, I can try open it manually and save it when I am back to my desk and see if it helps. And the question again, where is the setting in VS for this, so next time it will generate my changes this way?

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.cs (hunk 1 above)
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/Application.cs]: yes, we have to document public APIs

File: src/System.Windows.Forms.Design/src/Resources/SR.resx (hunk 6 above)
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms.Design/src/Resources/SR.resx]: I see, we don't have appropriate editorconfig settings for resx files:
https://github.com/dotnet/winforms/blob/f753f5bad925233c65a522921f9eeaf2b46b9c78/.editorconfig#L16

I think these 2 lines:

> charset = utf-8-bom
> insert_final_newline = true

Should be added to the resx section in the end of .editorconfig file as well.

File: src/System.Windows.Forms.Design/src/Resources/SR.resx (hunk 7 above)
Comment (zsd4yr) [reply src/System.Windows.Forms.Design/src/Resources/SR.resx]: my opinion would be to leave it then. Why bother every time?

File: src/System.Windows.Forms.Design/src/Resources/SR.resx (hunk 6 above)
Comment (zsd4yr) [reply src/System.Windows.Forms.Design/src/Resources/SR.resx]: they are specifed for only `.cs` files rn; I agree the rule should be extended to `.resx`

File: src/System.Windows.Forms.Design/src/Resources/SR.resx (hunk 7 above)
Comment (JuditRose) [reply src/System.Windows.Forms.Design/src/Resources/SR.resx]: I agree with @zsd4yr

File: .editorconfig [hunk 10]
```diff
@@ -161,8 +161,8 @@ indent_size = 2
 [*.{props,targets,config,nuspec}]
 indent_size = 2
 
-# resx Files
-[=.resx]
+# resx files
+[*.resx]
```
Comment (zsd4yr) [inline .editorconfig]: Could we also make this change as part of a stand-alone PR?

Comment (Tanya-Solyanik) [reply .editorconfig]: Resx extension is already mentioned on line 157

File: src/System.Windows.Forms.Design/src/Resources/xlf/SR.cs.xlf (hunk 8 above)
Comment (JuditRose) [reply src/System.Windows.Forms.Design/src/Resources/xlf/SR.cs.xlf]: I opened a separate PR to address resx  formatting settings: #548

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.cs [hunk 11]
```diff
@@ -709,12 +709,13 @@ internal static void OpenFormsInternalRemove(Form form)
             }
         }
 
-        internal static bool UseVisualStyles {
+#if (WINFORMS_NAMESPACE || WINFORMS_DESIGN_NAMESPACE)
+        public static bool UseVisualStyles {
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/Application.cs]: The reason to comment out this method was to keep it internal. Probably references to this method should be kept out as well? On desktop, .design namespace does not use this API, why do we need to include it in Core?

Comment (JuditRose) [reply src/System.Windows.Forms/src/System/Windows/Forms/Application.cs]: The only method referencing UseVisualStyles is in UnsafeNativeMethods, Activate, that method was public already. The issue surfaced when I moved these 2 files out from S.W.F. into the common folder

File: src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj [hunk 12]
```diff
@@ -4,6 +4,10 @@
     <TargetFramework>netcoreapp3.0</TargetFramework>
     <AssemblyName>System.Windows.Forms.Design</AssemblyName>    
     <CLSCompliant>false</CLSCompliant>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <NoWarn>$(NoWarn);618</NoWarn>
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj]: What is this error you are now suppressing? Why must it be suppressed?

File: src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj [hunk 13]
```diff
@@ -4,6 +4,10 @@
     <TargetFramework>netcoreapp3.0</TargetFramework>
     <AssemblyName>System.Windows.Forms.Design</AssemblyName>    
     <CLSCompliant>false</CLSCompliant>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj]: What does this do? Why is it necessary?

File: src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj [hunk 14]
```diff
@@ -21,4 +25,18 @@
     <ProjectReference Include="..\..\System.Windows.Forms\src\System.Windows.Forms.csproj" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Compile Include="..\..\Common\src\ApiHelper.cs" Link="Common\ApiHelper.cs" />
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj]: Is this currently everything in Common\src?  Should we consider including everything in Common\src?

File: src/System.Windows.Forms/src/System.Windows.Forms.csproj [hunk 15]
```diff
@@ -23,8 +23,18 @@
   <ItemGroup>
     <Compile Include="..\..\Common\src\ApiHelper.cs" Link="Common\ApiHelper.cs" />
```
Comment (zsd4yr) [inline src/System.Windows.Forms/src/System.Windows.Forms.csproj]: Same question, should we include all Common\src?

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.cs (hunk 11 above)
Comment (zsd4yr) [inline src/System.Windows.Forms/src/System/Windows/Forms/Application.cs]: Why are we excluding this method to these namespaces?

File: src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj (hunk 12 above)
Comment (JuditRose) [reply src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj]: This warn is coming from one of the files I moved to the common folder and was suppressed in System.Windows.Forms already.

File: src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj (hunk 13 above)
Comment (JuditRose) [reply src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj]: Same here, when I moved the files into the common folder I got unsafe block warnings during build and I saw that System.Windows.Forms has this line in the csproj file to suppress it, so I copied it over

File: src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj (hunk 14 above)
Comment (JuditRose) [reply src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj]: I didn't want to move files that has no reference elsewhere. We could do it in a separate PR, I would not make this one longer if not necessary.

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.cs (hunk 11 above)
Comment (JuditRose) [reply src/System.Windows.Forms/src/System/Windows/Forms/Application.cs]: https://github.com/dotnet/winforms/pull/480#discussion_r261852764

File: src/System.Windows.Forms/src/System.Windows.Forms.csproj (hunk 15 above)
Comment (JuditRose) [reply src/System.Windows.Forms/src/System.Windows.Forms.csproj]: I didn't want to reference the unused ones. It would only increase the size of the compiled code.

File: src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj (hunk 14 above)
Comment (zsd4yr) [reply src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj]: Is this currently everything in Common\src?

File: src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj (hunk 13 above)
Comment (zsd4yr) [reply src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj]: @Tanya-Solyanik @dreddy-work do you know what this is for?

File: src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj (hunk 12 above)
Comment (zsd4yr) [reply src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj]: I feel like we should add comments about _why_ these are being no warn'd

Comment (JuditRose) [reply src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj]: The original commit in System.Windows.Forms.csproj (this is where I copied this line over) is by dotnet bot with no explanation

Comment (JuditRose) [reply src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj]: Does anyone know?

Comment (ericstj) [reply src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj]: Even better: use pragmas instead and eliminate all NoWarns unless you really need to disable globally.  I believe 0618 lets you use obsolete types.   https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs0618

I believe WinForms was doing that because the obsolete usage is in public surface which can't be removed.

Comment (JuditRose) [reply src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj]: Thanks @ericstj !

Comment (zsd4yr) [reply src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj]: would you mind adding a comment in the file for this?

File: src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj (hunk 14 above)
Comment (zsd4yr) [reply src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj]: perfect, then explicit is probably the way to go. No need to over-engineer another solution

File: src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj (hunk 12 above)
Comment (JuditRose) [reply src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj]: I already have

File: src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj (hunk 13 above)
Comment (dreddy-work) [reply src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj]: We have unsafeblock in the code. @JuditRose Can you check if any of this ported code has any "unsafe" keywords used? if yes, we need this flag. otherwise, you may be able remove it.

Comment (JuditRose) [reply src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj]: Yes, I needed this to be able to build

File: src/Common/src/DpiHelper.cs [hunk 16]
```diff
@@ -8,14 +8,15 @@
 
 #if WINFORMS_NAMESPACE
 using CAPS = System.Windows.Forms.NativeMethods;
+#elif WINFORMS_DESIGN_NAMESPACE
+using CAPS = System.Windows.Forms.NativeMethods;
 #elif DRAWING_NAMESPACE
 using CAPS = System.Drawing.SafeNativeMethods;
 #elif DRAWINGDESIGN_NAMESPACE
```
Comment (zsd4yr) [inline src/Common/src/DpiHelper.cs]: Would you mind creating an issue to unify these namespace aliases? In particular, this one should be `DRAWING_DESIGN_NAMESPACE`

**Comment by zsd4yr** [comment]
One issue creation request; otherwise LGTM 😄 great stuff @JuditRose

File: src/Common/src/DpiHelper.cs (hunk 16 above)
Comment (JuditRose) [reply src/Common/src/DpiHelper.cs]: https://github.com/dotnet/winforms/issues/595

Comment (zsd4yr) [reply src/Common/src/DpiHelper.cs]: thx
